src/N3Reasoner.js

```diff
@@ -0,0 +1,194 @@
+import { termToId } from './N3DataFactory';
+
+export default class N3Reasoner {
+  constructor(store) {
+    this._store = store;
+  }
+
+  _add(subject, predicate, object, graphItem, cb) {
+    // Only add to the remaining indexes if there is not already a value in the index
+    if (!this._store._addToIndex(graphItem.subjects,   subject,   predicate, object)) return;
+    this._store._addToIndex(graphItem.predicates, predicate, object,    subject);
+    this._store._addToIndex(graphItem.objects,    object,    subject,   predicate);
+    cb();
+  }
+
+  // eslint-disable-next-line no-warning-comments
+  // TODO [FUTURE]: Improve performance by 'pre-computing' the index lookup,
+  // e.g., if a rule only has one variable, then we can just give it a pointer
+  // to the index that it should do lookups from.
+  // Similarly with insertions.
+  _evaluatePremise(rule, content, cb, i = 0) {
+    let v1, v2, value, index1, index2;
+    const [val0, val1, val2] = rule.premise[i].value, index = content[rule.premise[i].content];
+    const v0 = !(value = val0.value);
+    for (value in v0 ? index : { [value]: index[value] }) {
+      if (index1 = index[value]) {
```

Comment:
Note for self when I get around to finish cleaning this up. Due to the way that N3.js handles deletions; this is only here for the case where value is *defined* and !(value in index). So we can reduce a bunch of unnecessary calcs by moving this check outside of the loop.

One option is
```ts
for (value in v0 ? index : value in index ? { [value]: index[value] } : {})
   ...
```
alternatively